packages/database/src/migrations/20210201040026-AddDataSourceIdNotNullConstraints-modifies-schema.js

```diff
@@ -0,0 +1,39 @@
+'use strict';
+
+import { arrayToDbString } from '../utilities';
+
+var dbm;
+var type;
+var seed;
+
+/**
+ * We receive the dbmigrate dependency from dbmigrate initially.
+ * This enables us to not have to rely on NODE_PATH.
+ */
+exports.setup = function (options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+const NON_DATA_ELEMENT_QUESTION_TYPES = ['Instruction', 'PrimaryEntity', 'SubmissionDate'];
+
+exports.up = async function (db) {
+  await db.runSql(`ALTER TABLE survey ALTER COLUMN data_source_id SET NOT NULL;`);
+  await db.runSql(
+    `ALTER TABLE question
+    ADD CONSTRAINT data_source_id_not_null_on_conditions
+    CHECK(
+      type IN (${arrayToDbString(NON_DATA_ELEMENT_QUESTION_TYPES)}) OR
+      (data_source_id IS NOT NULL)
```

Comment:
Woah! Didn't realise constraints could get complex like this - that's awesome

Reply:
Yeah I was also excited to find out! 🚀

Reply:
One issue is that devs will have to remember updating this constraint when a new item is added in `NON_DATA_ELEMENT_QUESTION_TYPES`. I guess that' fine since this is an extra safety net- our code should not depend on it!